internal/gateway/kyma_gateway.go

```diff
@@ -0,0 +1,106 @@
+package gateway
+
+import (
+	"context"
+	_ "embed"
+	"fmt"
+	"github.com/kyma-project/api-gateway/apis/gateway/v1beta1"
+	"github.com/kyma-project/api-gateway/apis/operator/v1alpha1"
+	"github.com/kyma-project/api-gateway/controllers"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	nonGardenerDomainName = "local.kyma.dev"
+)
+
+// ReconcileKymaGateway reconciles the kyma-gateway and creates all required resources for the Gateway to fully work. It also adds a finalizer to
+// APIGateway CR and handles the deletion of the resources if the APIGateway CR is deleted.
+// Returns a Status object with the result of the reconciliation and an error if the reconciliation failed.
+func ReconcileKymaGateway(ctx context.Context, k8sClient client.Client, apiGatewayCR *v1alpha1.APIGateway) controllers.Status {
+
+	if isKymaGatewayEnabled(*apiGatewayCR) {
+		if !hasKymaGatewayFinalizer(*apiGatewayCR) {
+			if err := addKymaGatewayFinalizer(ctx, k8sClient, apiGatewayCR); err != nil {
+				return controllers.ErrorStatus(err, "Failed to add finalizer during Kyma Gateway reconciliation")
+			}
+		}
+	} else {
+		apiRuleExists, err := anyApiRuleExists(ctx, k8sClient)
```

Comment:
Shouldn't we always do it, not only in else? After all allowing deletion of the CR will allow deletion of the operator. And since the APIRule controller is in the same deployment, APIRules will no longer be reconciled + APIRule CRD will not be deleted

Reply:
For me this is a different level, because this is then bound to the Gateway CR and would be implemented on a different level. This is implemented in https://github.com/kyma-project/api-gateway/issues/559